Automattic/jetpack
# SHARING - Add general mobile Share button when Web Share is available

#### Changes proposed in this Pull Request:
* This PR adds a new client-generated "Share" button when social sharing is enabled and the current browser supports the [Web Share API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Share_API)
  - The injection of the button is dynamic because we can't necessarily tell (1) whether the browser supports the Web Share API, or (2) whether the current page can be shared -- I thought it better to add the button when needed, rather than having the button defined on the server and possibly flashing on the client before being removed.
  - In terms of styling, the new "Share" button uses the share icon from [social-logos](https://automattic.github.io/social-logos/) - this mirrors the Android share icon, so we may want to add the `share-ios` icon from [Gridicons](https://automattic.github.io/gridicons/) to `social-logos`.
  - Clicking on the button launches the native sharing menu with the post's permalink
    * We can consider adding further details, but I decided to stick with just the URL for this first iteration

#### Questions

* How should the presence of the button be controlled? Should it be on by default? Off by default?
* Should we show the iOS sharing icon to iOS users? (We'd need some way to detect when we're on iOS, and should be able to use the appropriate icon)
* Should we track how often this type of sharing is made available? Or only when it's triggered? (Given the API, detecting actual usage may be hard to determine correctly -- I think we can only detect whether the share menu was opened/launched.)

##### Screenshots

###### With text and icon
<img width="361" alt="jetpack-web-share-button" src="https://user-images.githubusercontent.com/3376401/169846850-c11db0f9-f6fb-4ba5-8499-621ccd4eaf15.png">

###### With only icon
<img width="361" alt="jetpack-web-share-icon" src="https://user-images.githubusercontent.com/3376401/169848556-d1bca411-094b-4029-8cf7-44c683393445.png">

**Related issue**: #16137

#### Jetpack product discussion

Initial discussion: p1HpG7-fLv-p2#sidebar-sharing-via-native-mobile-options
Dedicated discussion: p1HpG7-g44-p2

#### Does this pull request change what data or activity we track or use?

No, this PR does not change what data or activity we track.

#### Testing instructions:

* Ensure that you have a Jetpack test site with at least one post or page, and your test site is running the branch for this PR using Jetpack Beta.
* Navigate to Jetpack -> Settings, and then click on the Sharing tab
* In the "Sharing buttons" section, ensure that the "Add sharing buttons to your posts and pages" toggle is enabled
* Click on the "Configure your sharing buttons" link at the bottom of the screen
* On this screen, ensure that at least one sharing button is enabled, and that Posts and/or Pages will show the sharing buttons
* Click on the 'Edit "More" buttons' button, and ensure at least one button appears in this list
* Save any changes you made to this screen, and stay on this screen.
* In a separate tab, navigate to a page or post for the site
* Your primary browser should _not_ show the "Share" button
* Click on the "More" button, and verify that you don't see the generic Share button in the available options
* Close the "More" button popover
* Use a mobile browser or emulator that supports enabling Web Share (on my Mac, Safari emulates this correctly when picking an iOS device), and open the page for your site
* Verify that you see a Share button appear first in the list of sharing options (the actual visual will depend on the configured icon style for the sharing buttons)
* Click on the "More" button, and verify that you don't see the generic Share button in the available options
* Close the "More" button popover
* Click on the Share button, and verify that you see a share menu for your device with the URL of the post included in whatever target you select
* In your configuration window, change the "Button style" setting and save the changes
* Reload your mobile browser page/post, and verify that the Share button matches the styles of the other buttons (with the exception of the "Official" style, where we show the text and icon)
* Repeat the two steps above to verify that we handle the various button styles appropriately
